src/main/java/com/proofpoint/discovery/IpHostnameAuthManager.java

```diff
@@ -0,0 +1,190 @@
+                Map<String, String> properties = newAnnouncement.getProperties();
+                if (invalidHostnameIp(properties.get("http"), requesterAddress) ||
+                        invalidHostnameIp(properties.get("https"), requesterAddress) ||
+                        invalidHostnameIp(properties.get("admin"), requesterAddress)) {
+                    //Service is announcing a host with a different set of IPs
+                    throw new ForbiddenException();
+                }
+            }
+        }
+        catch (UnknownHostException e) {
+            //Unable to validate an IP or look up a hostname
+            auditLogger.audit(auditRecord(e.getMessage()));
+            throw new ForbiddenException();
+        }
+    }
+
+    @Override
+    public void checkAuthDelete(Id<Node> nodeId, HttpServletRequest request)
+    {
+        try {
+            InetAddress requesterAddress = InetAddress.getByName(request.getRemoteAddr());
+            String nodeAnnouncer = dynamicStore.getAnnouncer(nodeId);
+            if (mismatchedAnnouncer(nodeAnnouncer, requesterAddress)) {
+                auditLogger.audit(auditRecord("IP %s tried to delete node %s owned by IP %s", requesterAddress.getHostAddress(), nodeId.toString(), nodeAnnouncer));
+                throw new ForbiddenException();
+            }
+        }
+        catch (UnknownHostException e) {
+            auditLogger.audit(auditRecord(e.getMessage()));
+            throw new ForbiddenException();
+        }
+    }
+
+    @Override
+    public void checkAuthReplicate(HttpServletRequest request)
+    {
+        try {
+            InetAddress requesterAddress = InetAddress.getByName(request.getRemoteAddr());
+            if (!discoveryAddrs.contains(requesterAddress)) {
+                auditLogger.audit(auditRecord("IP %s tried to replicate as Discovery", requesterAddress.getHostAddress()));
+                throw new ForbiddenException();
+            }
+        }
+        catch (UnknownHostException e) {
+            auditLogger.audit(auditRecord(e.getMessage()));
+            throw new ForbiddenException();
+        }
+    }
+
+    private boolean mismatchedAnnouncer(@Nullable String nodeAnnouncer, InetAddress requesterAddr)
+            throws UnknownHostException
+    {
+        return nodeAnnouncer != null && !requesterAddr.equals(InetAddress.getByName(nodeAnnouncer));
+    }
+
+    private boolean invalidHostnameIp(String hostname, InetAddress requesterAddress)
+    {
+        if (hostname == null) {
+            return false;
+        }
+        try {
+            URI hostUri = URI.create(hostname);
+            InetAddress[] hostIps = InetAddress.getAllByName(hostUri.getHost());
+            if (!Arrays.asList(hostIps).contains(requesterAddress)) {
+                auditLogger.audit(auditRecord("IP %s tried to announce other host %s", requesterAddress.getHostAddress(), hostUri.getHost()));
```

Comment:
This is sort of missing the point of logging in JSON. The requester IP and host should be separate attributes in the record, so people and/or code that is reviewing the log can easily search for or generate statistics from them.

It looks like there are four or five conditions being logged, which should be as many log record types:

* IP X tried to announce other host Y
* IP X tried to Y [update/delete] node Z owned by IP A (4 attributes)
* IP X tried to replicate as discovery
* DNS lookup on announced host X failed with Y
* DNS lookup while parsing an IP address failed (could be avoided by using `InetAddresses.forString()`)